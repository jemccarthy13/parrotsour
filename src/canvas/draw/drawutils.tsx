/**
 * This file contains common low-level drawing utilities used
 * to build groups and pictures.
 */

// Interfaces
import { Point } from "../../classes/point"

/**
 * 'Clamp' the location to the confines of the drawing context
 * @param ctx The context to constrict it to
 * @param pos the current position
 */
export function clampInContext(
  ctx: CanvasRenderingContext2D,
  x: number | Point,
  y?: number
): Point {
  let retPoint = Point.DEFAULT
  if (x instanceof Point) {
    retPoint = new Point(
      Math.min(Math.max(x.x, 0), ctx.canvas.width),
      Math.min(Math.max(x.y, 0), ctx.canvas.height)
    )
  } else {
    if (y === undefined) y = 0
    retPoint = new Point(
      Math.min(Math.max(x, 0), ctx.canvas.width),
      Math.min(Math.max(y, 0), ctx.canvas.height)
    )
  }
  return retPoint
}
