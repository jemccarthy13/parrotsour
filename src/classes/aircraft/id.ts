export enum IDMatrix {
  HOSTILE = "red",
  SUSPECT = "orange",
  ASSUME_FRIEND = "green",
  NEUTRAL = "purple",
  FRIEND = "blue",
}

export function getMostRestrictiveID(ids: IDMatrix[]): IDMatrix {
  function idToNum(id: IDMatrix) {
    const num = 4
    switch (id) {
      case IDMatrix.SUSPECT:
        return 3
      case IDMatrix.NEUTRAL:
        return 2
      case IDMatrix.ASSUME_FRIEND:
        return 1
      case IDMatrix.FRIEND:
        return 1
    }
    return num
  }
  let lowestID = IDMatrix.HOSTILE
  let lowestIDNum = 4
  for (let i = 0; i < ids.length; i++) {
    const idNum = idToNum(ids[i])
    if (idNum < lowestIDNum) {
      lowestIDNum = idNum
      lowestID = ids[i]
    }
  }
  return lowestID
}
