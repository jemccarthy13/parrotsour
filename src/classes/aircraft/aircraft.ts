// Classes and interfaces
import { DataTrail } from "./datatrail/datatrail"
import { SensorType } from "./datatrail/sensortype"
import { GroupParams } from "../groups/group"
import { IDMatrix } from "./id"
import { AircraftIntent, IntentParams } from "./intent"
import { Point } from "../point"
import Tasking from "../taskings/tasking"

// Functions
import {
  getDegDeltaBetween,
  headingToRadians,
  randomNumber,
} from "../../utils/psmath"
import { DataTrailFactory } from "./datatrail/datatrailfactory"

/*
 * TODO -- EWI -- for fill-ins, implement more AC types?
 */
export enum ACType {
  FTR,
  RPA,
  TANK,
  AEW,
}

/**
 * Aircraft have all the parameters of Groups, plus a single
 * altitude and configurable type.
 */
interface AircraftParams extends GroupParams {
  alt: number
  type: ACType
}

/**
 * TODO -- AIRCRAFT -- consider moving to a structure like
 * https://khalilstemmler.com/blogs/typescript/getters-and-setters/
 * To reduce the number of functions provided in the aircraft "api"
 */
export class Aircraft {
  // AASI (alt, aspect, speed, ID)
  private altitude: number
  private heading: number
  // TODO -- SPEED -- implement aircraft speed logic
  // private speed: number
  private id: IDMatrix

  private dataTrail: Map<number, DataTrail>

  private type: ACType

  private startPos = new Point(0, 0)

  private intent = new AircraftIntent()
  private tasking: Tasking | undefined
  private capping = false

  private ctx: CanvasRenderingContext2D | undefined

  constructor(p?: Partial<AircraftParams>) {
    if (!p) p = {}

    // Take given heading/ID/EWI or use default
    this.heading = p.hdg || 90
    this.id = p.id || IDMatrix.HOSTILE
    this.type = p.type || ACType.FTR

    // Set current position
    this.startPos.x = p.sx || randomNumber(1, 100)
    this.startPos.y = p.sy || randomNumber(1, 100)

    // TODO -- DATATRAIL -- store a map of SensorType -> DataTrail
    // here, so it has both kinds of data trails available.
    // Then, in the PSControls, toggling data trail types only
    // triggers a redraw of existing groups' data trails (via .draw())
    this.dataTrail = new Map<number, DataTrail>()

    for (const type in SensorType) {
      if (!Number.isNaN(Number(type)))
        this.dataTrail.set(
          Number(type),
          DataTrailFactory.create(
            Number(type),
            this.startPos,
            this.getHeading()
          )
        )
    }

    // Set current altitude
    let low = 15
    let hi = 45
    if (this.type === ACType.RPA) {
      low = 0o5
      hi = 18
    }
    this.altitude = p.alt || randomNumber(low, hi)

    // Current matches desired initially
    this.intent.setDesiredHeading(this.heading)
  }

  /*************************************************************************
   * Location
   *************************************************************************/
  /**
   * @returns Start position for where this Aircraft's draw will start; will be approximately
   * the 'rear'' end of the data trail.
   */
  public getStartPos(): Point {
    return this.startPos
  }

  /**
   * Get the center of mass position for this group.
   *
   * The type of data trail dictates where the center of mass is.
   * (i.e. radar - last sensed radar vs. arrow - loc of arrow head)
   *
   * @param dataStyle (Radar & IFF) | Arrows - must explicitly be set if drawing
   *                  anything other than arrows
   * @param ctx Current drawing context
   */
  getCenterOfMass(dataStyle?: SensorType): Point {
    if (this.isCapping()) {
      return this.getStartPos()
    }
    dataStyle = dataStyle === undefined ? SensorType.ARROW : dataStyle
    const dataTrail = this.dataTrail.get(dataStyle)
    if (dataTrail) {
      return dataTrail.getCenterOfMass(this.getHeading())
    } else {
      return new Point(-1, -1)
    }
  }

  /**
   * @returns true iff Aircraft is capping (at desired final Route location)
   */
  isCapping(): boolean {
    return this.capping
  }

  /**
   * @param newVal New value for isCapping
   */
  setCapping(newVal: boolean): void {
    this.capping = newVal
  }

  /*************************************************************************
   * Heading and track direction/aspect
   *************************************************************************/
  /**
   * @returns Current heading of this Aircraft
   */
  public getHeading(): number {
    return this.heading
  }

  /**
   * @param hdg Set new heading of this Aircraft
   */
  public setHeading(hdg: number): void {
    this.heading = hdg
  }

  /*************************************************************************
   * Altitude
   *************************************************************************/
  /**
   * @returns Current altitude of this Aircraft
   */
  public getAltitude(): number {
    return this.altitude
  }

  /*************************************************************************
   * Speed -- TODO
   *************************************************************************/

  /*************************************************************************
   * ID
   *************************************************************************/
  /**
   * @returns Current IDMatrix (HOS/NEU/etc) of this Aircraft.
   */
  getIDMatrix(): IDMatrix {
    return this.id
  }
  /**
   * @param newID Set new IDMatrix ID for this Aircraft
   */
  setIDMatrix(newID: IDMatrix): void {
    this.id = newID
  }

  /*************************************************************************
   * Fillins
   *************************************************************************/
  /**
   * @returns Current Aircraft type
   */
  getType(): ACType {
    return this.type
  }

  /*************************************************************************
   * Drawing and animation
   *************************************************************************/
  /**
   * Draw the Aircraft, with it's associated data trail and symbology
   * @param context  Current drawing style
   * @param dataStyle The SensorType of the DataTrail
   */
  draw(dataStyle: SensorType): void {
    const dataTrail = this.dataTrail.get(dataStyle)
    if (dataTrail) {
      dataTrail.draw(this.getHeading(), this.getIDMatrix())
    }
  }

  /**
   * Move the arrow once based on the current heading / vector.
   */
  move(): void {
    if (!this.capping) {
      // don't move if capping at final destination
      // calculate & apply offsets based on start/end position
      const rads: number = headingToRadians(this.getHeading()).radians
      const offsetX: number = 7 * Math.cos(rads)
      const offsetY: number = -7 * Math.sin(rads)
      this.startPos.x += offsetX
      this.startPos.y += offsetY

      this.dataTrail.forEach((value: DataTrail) => {
        value.move(this.getHeading())
      })

      // apply a turn as required
      this.turnToTarget()
    }
  }

  /**
   * Make a logical turn towards the Aircraft's target by applying an offset
   * to the current heading.
   */
  turnToTarget(): void {
    // if Aircraft wants to go somewhere
    // determine the bearing to the next route point
    const tgtPos = this.intent.getNextRoutingPoint()
    if (tgtPos) {
      this.intent.setDesiredHeading(
        this.getCenterOfMass().getBR(tgtPos).bearingNum
      )
    }

    // how "far away" from current heading is the desired?
    const turnDegrees = getDegDeltaBetween(
      this.getHeading(),
      this.intent.getDesiredHeading(),
      this.intent.getForcedTurn()
    )

    // apply 'logical' turn by only turning so far towards the target
    // each frame
    let divisor = 7
    const absDelt = Math.abs(turnDegrees)
    if (absDelt >= 90) {
      divisor = 15
    } else if (absDelt < 7) {
      divisor = 1
    }

    this.setHeading(Math.floor(this.getHeading() + turnDegrees / divisor))
  }

  /**
   * If not at desired altitude, increment to climb/descend towards intended altitude.
   */
  doNextAltChange(): void {
    const atDesiredAlt = this.intent.getDesiredAltitude() === this.getAltitude()
    const CHANGE_PER_FRAME = 0.5
    if (!atDesiredAlt) {
      if (this.intent.getDesiredAltitude() > this.getAltitude()) {
        this.altitude += CHANGE_PER_FRAME // climb
      } else {
        this.altitude -= CHANGE_PER_FRAME // descend
      }
    }
  }

  /*************************************************************************
   * Routing & Intent
   *************************************************************************/
  /**
   *
   * @param newIntent Update this aircraft's intent based on Params
   */
  updateIntent(newIntent: Partial<IntentParams>): void {
    this.intent.updateIntent(newIntent)
  }

  /**
   * @returns true iff there are no desired locations left in the intended route
   */
  atFinalDestination(): boolean {
    return this.intent.atFinalDestination()
  }

  /**
   * @returns Point, if there's a next routing point, or
   * undefined if no intended destination set
   */
  getNextRoutingPoint(): Point | undefined {
    return this.intent.getNextRoutingPoint()
  }

  /**
   * @param pt A new point to be added to the Aircraft's route. It is added to the end
   * of the list.
   */
  addRoutingPoint(pt: Point): void {
    this.intent.addRoutingPoint(pt)
  }

  /**
   * Check current against intended, modify heading and intent as necessary.
   */
  doNextRouting(): void {
    if (!this.intent.atFinalDestination()) {
      // Aircraft still has a route but we've reached our next destination
      // so pop the first point out of the intended route
      if (this.intent.atNextRoutingPoint(this.getStartPos())) {
        this.intent.removeRoutingPoint()
      }
      // If Aircraft still has a point left in routing, update desired heading
      // and turn towards it
      const nextPt = this.intent.getNextRoutingPoint()
      if (nextPt)
        this.updateIntent({
          desiredHeading: this.getCenterOfMass().getBR(nextPt).bearingNum,
        })
      this.turnToTarget()
    } else {
      // Aircraft has reached final destination, so if intent matches it is capping
      this.capping = this.intent.atFinalDestination()
    }
  }

  clearRouting(): void {
    this.intent.clearRouting()
  }

  /*************************************************************************
   * Taskings
   *************************************************************************/
  /**
   * @returns true iff Aircraft has a tasking assigned.
   */
  isTasked(): boolean {
    return this.tasking !== undefined
  }

  /**
   * Clears the current tasking.
   */
  clearTasking(): void {
    this.tasking = undefined
  }

  /**
   * @param task Set a new Tasking for this Aircraft.
   */
  setTasking(task: Tasking): void {
    this.tasking = task
  }

  /**
   * @returns The current Tasking, or undefined if Aircraft has none.
   */
  getTasking(): Tasking | undefined {
    return this.tasking
  }
}
