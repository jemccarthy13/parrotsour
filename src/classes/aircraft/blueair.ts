import { AircraftGroup } from "../groups/group"

export class BlueAir {
  private static blue: AircraftGroup

  public static set(grp: AircraftGroup): void {
    this.blue = grp
  }

  public static get(): AircraftGroup {
    return this.blue
  }
}
