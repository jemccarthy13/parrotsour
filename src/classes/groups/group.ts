// Classes & Interfaces
import { SensorType } from "../aircraft/datatrail/sensortype"
import { Point } from "../point"
import { ACType, Aircraft } from "../aircraft/aircraft"
import { getMostRestrictiveID, IDMatrix } from "../aircraft/id"
import { IntentParams } from "../aircraft/intent"
import Tasking from "../taskings/tasking"
import { FORMAT } from "../supportedformats"

import { AltStack, getAltStack } from "../altstack"
// Functions
import {
  headingToRadians,
  PIXELS_TO_NM,
  randomNumber,
} from "../../utils/psmath"
import { drawGroupCap } from "./groupcap"
import { Braaseye } from "../braaseye"
import { Aspect, aspectFromCATA, toCardinal } from "../../utils/aspect"
import { BRAA } from "../braa"
import RangeBack from "./rangeback"

/**
 * The types of data that can be used to seed a group.
 *
 * Everything not provided is randomized or defaulted.
 */
export interface GroupParams {
  dataTrailType: SensorType
  sx?: number
  sy?: number
  nContacts?: number
  hdg?: number
  alts?: number[]
  desiredHdg?: number
  id?: IDMatrix
  type?: ACType
}

/**
 * An AircraftGroup is an array of 1+ Aircraft.
 *
 * It contains logic to move, give intent to, and draw Aircraft.
 *
 * TODO -- GROUP SPLIT/MERGE -- Currently maneuvers are handled as an all
 * or nothing. In the animator, consider additional logic for (man > 0), to update
 * "groups" var to split groups. Should be 'easy'
 */
export class AircraftGroup extends Array<Aircraft> {
  private braaseye!: Braaseye

  private startPos: Point = Point.DEFAULT
  private label = "GROUP"
  private useTrackDir = true
  private maneuvers = 0
  private anchor = false
  private useBull = false

  /**
   * Construct an AircraftGroup from the given parameters.Group will initialize
   * what it can, randomize what it can, and default the rest.
   *
   * @param p A Partial object of GroupParams.
   */
  constructor(p?: Partial<GroupParams>) {
    super()
    if (!p) p = {}
    this.startPos.x = p.sx || randomNumber(1, 100)
    this.startPos.y = p.sy || randomNumber(1, 100)

    // Seed params to pass into Aircraft
    p.sx = this.startPos.x
    p.sy = this.startPos.y
    p.hdg = p.hdg || 90

    this.startPos = new Point(p.sx, p.sy)

    // Create nContacts number of Aircraft and add to this collection
    const nContacts = p.nContacts || randomNumber(1, 4)
    for (let contact = 0; contact < nContacts; contact++) {
      if (p.alts && p.alts[contact])
        this.push(new Aircraft({ ...p, alt: p.alts[contact] }))
      else {
        this.push(new Aircraft(p))
      }

      // Compute start position offset for follow-on groups (calculated from
      // 90 deg perpendicular)
      const vectors = headingToRadians(p.hdg)
      p.sx += 2 * PIXELS_TO_NM * Math.cos(vectors.offset)
      p.sy += 2 * PIXELS_TO_NM * -Math.sin(vectors.offset)
    }

    // TODO -- MANEUVER -- % chance of 2 maneuvers (i.e. flank turn back hot)
    if (randomNumber(0, 100) < 20) {
      this.maneuvers = 1
    }
  }

  /*************************************************************************
   * Formatting
   *************************************************************************/
  setAnchor(b: boolean): void {
    this.anchor = b
    this.useBull = b
  }

  isAnchor(): boolean {
    return this.anchor
  }

  setUseBull(b: boolean): void {
    this.useBull = b
  }

  getUseBull(): boolean {
    return this.useBull
  }

  formatNumContacts(): string {
    let answer = ""
    if (this.getStrength() > 1) {
      if (this.getStrength() >= 3) {
        answer += "HEAVY "
      }
      answer += this.getStrength() + " CONTACTS"
    }
    return answer
  }

  /**
   * Return the string formatted answer for this group based on properties of the group
   */
  format(format: FORMAT, rangeBack?: RangeBack): string {
    // format label
    let answer = this.getLabel() + " "

    // format separation
    if (rangeBack) {
      answer += rangeBack.toString() + " "
    }

    // format bullseye if anchor priority
    const braaseye = this.getBraaseye()
    if (this.useBull || false) {
      answer += "BULLSEYE " + braaseye.bull.toString() + ", "
    }

    // format altitude stack
    const altStack = this.getAltStack(format)
    answer += altStack.stack + " "

    // format track direction
    const trackDir = this.getTrackDir()
    answer += " "
    answer += trackDir !== undefined ? trackDir : ""

    // apply ID
    answer += " HOSTILE "

    // apply fill-in for # contacts
    answer += this.formatNumContacts() + " "

    // apply fill-ins (HI/FAST/etc)
    answer += " " + altStack.fillIns
    return answer.replace(/ {2}/g, " ")
  }

  /*************************************************************************
   * Location
   *************************************************************************/
  /**
   * @returns Start position for where this group's draw will start; will be approximately
   * the 'rear' end of the data trail.
   */
  getStartPos(): Point {
    return this[0].getStartPos()
  }

  /**
   * "Center of Mass" for an aircraft/group is one projected data plot ahead of the
   * data trail, averaged. For Aircraft, this is equiv. to one projected data plot ahead
   * of the Aircraft, on it's current heading.
   *
   * @returns The "center of mass" for this group, which is the average of each
   * Aircraft's center of mass.
   */
  getCenterOfMass(dataStyle: SensorType): Point {
    if (this.isCapping()) {
      return this.startPos
    }
    let x = 0
    let y = 0
    for (let idx = 0; idx < this.getStrength(); idx++) {
      const acPos = this[idx].getCenterOfMass(dataStyle)
      x += acPos.x
      y += acPos.y
    }
    return new Point(x / this.getStrength(), y / this.getStrength())
  }

  /**
   * @returns true iff one Aircraft in this group is capping
   */
  isCapping(): boolean {
    return (
      this.find((ac) => {
        return !ac.isCapping()
      }) === undefined
    )
  }

  /**
   * Set each aircraft in this group to be capping
   */
  setCapping(newVal: boolean): void {
    this.forEach((ac) => ac.setCapping(newVal))
  }

  setBraaseye(braaseye: Braaseye): void {
    this.braaseye = braaseye
  }

  getBraaseye(): Braaseye {
    return this.braaseye
  }

  /*************************************************************************
   * Heading and track direction/aspect
   *************************************************************************/

  /**
   * Get 'aspect' (HOT/FLANK/BEAM, etc) to another group
   *
   * Aspect is calculated by taking the angle difference between
   * other a/c heading, and the reciprocal bearing between this group
   * and other a/c.
   *
   * @param otherGrp other aircraft
   */
  getAspect(otherGrp: AircraftGroup, dataStyle: SensorType): Aspect {
    const recipBrg: BRAA = otherGrp
      .getCenterOfMass(dataStyle)
      .getBR(this.getCenterOfMass(dataStyle))

    let dist = (otherGrp.getHeading() - parseInt(recipBrg.bearing) + 360) % 360
    if (dist > 180) dist = 360 - dist
    const cata = dist

    return aspectFromCATA(cata)
  }

  /**
   * @returns Current heading of the Group. Currently assumes all Aicraft are
   * following the same heading, therefore only returns first Aircraft's heading.
   */
  getHeading(): number {
    return this[0].getHeading()
  }

  /**
   * @param newHdg New heading to set for all aircraft in this group
   */
  setHeading(newHdg: number): void {
    this.forEach((ac) => ac.setHeading(newHdg))
  }

  /**
   * @returns Cardinal track direction or undefined. If there is a picture
   * track direction specified (i.e. all groups track West) this function
   * returns undefined. Otherwise, cardinal direction from heading.
   */
  getTrackDir(): string | undefined {
    let trackDir = undefined
    if (this.isCapping()) {
      trackDir = "CAP"
    } else if (this.useTrackDir) {
      trackDir = "TRACK " + toCardinal(this.getHeading())
    }
    return trackDir
  }

  /**
   * Set the useTrackDir flag, whether or not to include track direction
   * in group formatting
   */
  setUseTrackDir(newVal: boolean): void {
    this.useTrackDir = newVal
  }

  /**
   * Update maneuver count
   * @param newNumManeuvers New number of maneuvers remaining
   */
  setManeuvers(newNumManeuvers: number): void {
    this.maneuvers = newNumManeuvers
  }

  /**
   * @returns true iff group has at least one maneuver left.
   */
  doesManeuvers(): boolean {
    return this.maneuvers > 0
  }

  /*************************************************************************
   * Altitude(s)
   *************************************************************************/
  /**
   * @returns Altitude of the highest Aircraft in this group
   */
  getAltitude(): number {
    return Math.max(...this.getAltitudes())
  }

  /**
   * @param format The desired formatting
   * @returns AltStack containing formatted STACKS/Fillins
   */
  getAltStack(format: FORMAT): AltStack {
    return getAltStack(this.getAltitudes(), format)
  }

  /**
   * @returns number[] containing the altitude of every Aircraft in this Group
   */
  getAltitudes(): number[] {
    const alts = []
    for (let idx = 0; idx < this.length; idx++) {
      alts.push(this[idx].getAltitude())
    }
    return alts
  }

  /*************************************************************************
   * Speed -- TODO
   *************************************************************************/

  /*************************************************************************
   * ID
   *************************************************************************/
  /**
   * @returns Current group label
   */
  getLabel(): string {
    return this.label
  }

  /**
   * @param newLbl New group label
   */
  setLabel(newLbl: string): void {
    this.label = newLbl
  }

  /**
   * Update group ID
   * @param newID
   */
  setIDMatrix(newID: IDMatrix): void {
    this.forEach((ac) => {
      ac.setIDMatrix(newID)
    })
  }

  /**
   * @returns The most restrictive ID of child Aircraft
   */
  getIDMatrix(): IDMatrix {
    const ids: IDMatrix[] = []
    this.forEach((ac) => {
      ids.push(ac.getIDMatrix())
    })
    return getMostRestrictiveID(ids)
  }

  /*************************************************************************
   * Fillins
   *************************************************************************/
  /**
   * @returns Number of contacts in this group
   */
  getStrength(): number {
    return this.length
  }

  /**
   * @returns The Type of Aircraft in this group. Currently only supports one type
   * for the entire group (first aircraft)
   */
  getType(): ACType {
    return this[0].getType()
  }

  /*************************************************************************
   * Drawing and animation
   *************************************************************************/
  /**
   * Draw each aircraft. See Aircraft.draw(...)
   * @param context Current drawing context
   * @param dataStyle The type of DataTrail to use
   */
  draw(dataStyle: SensorType): void {
    if (this.isCapping()) {
      drawGroupCap(this)
    } else {
      this.forEach((ac) => ac.draw(dataStyle))
    }
  }

  /**
   * Move each Aircraft. See Aircraft.move()
   */
  move(): void {
    this.forEach((ac) => ac.move())
  }

  /**
   * Update each Aircraft's altitude based on intent. See Aircraft.doNextAltChange()
   */
  updateAltitude(): void {
    this.forEach((ac) => ac.doNextAltChange())
  }

  /*************************************************************************
   * Routing
   *************************************************************************/
  /**
   * Update intent for each Aircraft in the group. See Aircraft.updateIntent()
   * @param newIntent Object representing new intent for each Aircraft.
   */
  updateIntent(newIntent: Partial<IntentParams>): void {
    this.forEach((ac) => ac.updateIntent(newIntent))
  }

  /**
   * @returns true iff there is an Aircraft with intended destination
   */
  hasRouting(): boolean {
    return this.find((ac) => !ac.atFinalDestination()) !== undefined
  }

  /**
   * @returns The next Point on this group's intended route. Assumes
   * all Aircraft in the Group are following the same route, therefore returns
   * only the first Aircraft's next routing point.
   */
  getNextRoutingPoint(): Point | undefined {
    return this[0].getNextRoutingPoint()
  }

  /**
   * @param pt Point to add to all group Aircraft(s) routes
   */
  addRoutingPoint(pt: Point): void {
    this.forEach((ac) => ac.addRoutingPoint(pt))
  }

  /**
   * Perform the next routing action for each aircraft. See Aircraft.doNextRouting()
   */
  doNextRouting(): void {
    this.forEach((ac) => ac.doNextRouting())
  }

  /**
   * Clears all currently assigned transit locations
   */
  clearRouting(): void {
    this.forEach((ac) => ac.clearRouting())
  }

  /*************************************************************************
   * Tasking
   *************************************************************************/
  /**
   * @param task Aircraft's new Tasking.
   */
  setTasking(task: Tasking): void {
    this.forEach((ac: Aircraft) => {
      ac.setTasking(task)
    })
  }

  /**
   * @returns true if there is at least one Aircraft in this group that has a Tasking.
   */
  isOnTask(): boolean {
    return this.find((ac) => ac.isTasked()) !== undefined
  }
}
