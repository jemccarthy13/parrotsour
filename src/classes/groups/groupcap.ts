import { PaintBrush } from "../../canvas/draw/paintbrush"
import { AircraftGroup } from "./group"

/**
 * Draw a capping group
 *
 * TODO -- since it takes only group now, maybe move to group file?
 *
 * @param {AircraftGroup} group Group to draw capping
 */
export function drawGroupCap(group: AircraftGroup): void {
  const c = PaintBrush.getContext()

  const id = group.getIDMatrix()

  c.lineWidth = 1
  c.fillStyle = id
  c.strokeStyle = id

  c.beginPath()

  let radius = 10

  const startPos = group.getStartPos()
  const startX = startPos.x
  const startY = startPos.y

  const strength = group.getStrength()

  if (strength === 1) {
    c.arc(startX, startY, 10, 1.0 * Math.PI, 0.8 * Math.PI)
    c.stroke()
    PaintBrush.drawLine(startX - 8, startY + 6, startX - 6, startY + 12, id)
  } else {
    const ratio = 2 / strength - 0.1
    let startPI = 0
    let endPI = ratio
    radius = 12
    for (let x = 1; x <= strength; x++) {
      c.arc(startX, startY, radius, startPI * Math.PI, endPI * Math.PI)
      c.stroke()

      const opp: number = radius * Math.sin(endPI * Math.PI)
      const adj: number = radius * Math.cos(endPI * Math.PI)

      const endy = startY + opp
      const endx = startX + adj

      c.beginPath()
      c.moveTo(startX + adj * 0.6, startY + opp * 0.9)
      c.lineTo(endx, endy)
      c.stroke()
      c.beginPath()

      startPI = endPI + 0.1
      endPI = startPI + ratio
    }
  }
}
