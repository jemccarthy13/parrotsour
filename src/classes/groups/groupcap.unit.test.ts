import { PaintBrush } from "../../canvas/draw/paintbrush"
import { IDMatrix } from "../aircraft/id"
import { AircraftGroup } from "./group"
import { drawGroupCap } from "./groupcap"

// eslint-disable-next-line @typescript-eslint/no-var-requires
const canvasSerializer = require("jest-canvas-snapshot-serializer")
expect.addSnapshotSerializer(canvasSerializer)

describe("draw_group_cap", () => {
  const canvas = document.createElement("canvas")
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const ctx = canvas.getContext("2d")!
  PaintBrush.use(ctx)

  afterEach(() => {
    PaintBrush.clearCanvas()
  })

  it("draws_capping_group", () => {
    const grp = new AircraftGroup({
      nContacts: 4,
      hdg: 135,
      id: IDMatrix.HOSTILE,
      sx: 50,
      sy: 50,
    })
    grp.setCapping(true)
    drawGroupCap(grp)
    expect(canvas).toMatchSnapshot()
  })

  it("draws_singleship_cap", () => {
    const grp = new AircraftGroup({
      nContacts: 1,
      hdg: 135,
      id: IDMatrix.HOSTILE,
      sx: 50,
      sy: 50,
    })
    grp.setCapping(true)
    drawGroupCap(grp)
    expect(canvas).toMatchSnapshot()
  })
})
