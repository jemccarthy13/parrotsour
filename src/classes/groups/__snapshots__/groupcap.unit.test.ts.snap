// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`draw_group_cap draws_capping_group 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAAC/ElEQVR4nO3dv2uVVxgH8O+VCMHopMlQ3DJIhqaGVIrQ/6IEd9dMxdg5Y2Ps2DlGC536j5QW6aJTQUsGxRRashQqPA45lNsq2jTn5o29n890uT++PNPDec977vMmAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMcwGrqAXirZSPJrksdJfhwlfw5cEtDZmaELOKpKrlVyr5L7lSyNfbSQ5NMkXyfZr2S3kmvDVAlMtUoWKnlQyZNKblWyXsn39YamW8nFSjYqedoa28IQNQNTqJKPKvmlkq1Kzo29P/OO381Vst2a3PLkKwWmWiVXK3leyWfHyFir5JmmBUxMuwx8epxmNZa11lZa8z1qA/ibSr6pZKtj3nYlu73yAJIklXzSVkTn3v3tf50511Zsq70yAdLuCH7eOfNCJV9W8m3PXGCKVXK2kt8rudgha7GSvUr+qOSgrbA2O5QJkFRyvZIfOmXNVHK5ktkeecBw3nqOaUBLSX7qETRKXibZ65EFDOu0/jVnPsmLoYsAGFw73rAxdB3A0ZzWFdakXUqyP3QRwNG8Fw2rktm2cd5rz20lyaNOWcC0q2SzHUE4aEcS9ipZ7JA7X8lvHZsfcEJO8wrrSg4PeH4wSmZHyeVR8nOH3JtJvmt3DwGOr5KP2wprrmPm+TamZqVXJkCSpA3fu9Mx76tKdnrlAfxlbLzMWoesG8bLABNVyXIb4Pefm1ZrVs8q+bBnbQCvaU3rSTvw+dY9rfG7f22czN32W80KOBntOMJuu0S8XYeHP//5nTPtwRTrlXzRNtjvuQwEBlHJaiU7bfzMw/Z6e+zzpbZZv2NQH/y/vLcPUm2Xfqs5nOxwaZTcHbgkAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACASXoFwb7VzJzIti8AAAAASUVORK5CYII="
/>
`;

exports[`draw_group_cap draws_singleship_cap 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASwAAACWCAYAAABkW7XSAAAABmJLR0QA/wD/AP+gvaeTAAAEpUlEQVR4nO3cPajVdRzH8c+9OvSwOvhQ0SiJPWilFLo0GOoiSI22GBdEmsSh2Why0OgJAqcchIpIaJcQHzGKsCGiQkS4bU4WnoY/QUOBXs8953y6r9d44P/jO735/c/5/c7caDQaBaDA/LQHALhXggXUECyghmABNQQLqCFYQA3BAmoIFlBDsIAaggXUECyghmABNQQLqCFYQA3BAmoIFlBDsIAaggXUECyghmABNQQLqCFYQA3BAmoIFlBDsIAaggXUECyghmABNQQLqCFYQA3BAmoIFlBDsIAaggXUECyghmABNQQLqCFYQI3V0x7ggVy/nly8mCwuJn/+maxfnzz3XLJp07QnA5bB3Gg0Gk17iPty+3by8cfJiRPJ/Hzy0kvJ2rXJqlXJjRvJ+fPJ3bvJ4cPJwkLyyCPTnhgYk65gnTuXvPFG8sILydGjw27q31y7lrzzTnL5cvLpp8n27RMdE1gePcH67LPk0KHkk0+S3bvv7ZmvvkoOHkzefz/Zt2955wOWXUewvvkm2b8/+frr5Jln7u/Zb79Ndu1KzpxJduxYnvmAiZj9YN2+nWzenJw8mezdu7Q1zp5N3npriNejj453PmBiZv9Yw0cfJdu2LT1WSbJnT7J1a/Lhh+ObC5i42d9hPflk8sUXybPPPtg6ly4lr7+e/PRTMjc3ltGAyZrtHdYPPwzHFR40Vsnwy+LcXPLjjw++FjAVsx2sy5eHc1bjsn37cNAUqDTbwbp1K1m3bnzrrV2b/P77+NYDJmq2r+YcOTLe9VavTv74Y7xrAhMz2zuscfvll+SJJ6Y9BbBEKytYFy8mTz897SmAJVo5wbpyZfjF8amnpj0JsEQrJ1jHjg13EYFaHcG6c2e4T7hUX345XMtZWBjfTMDEdQTr4MFk587ku+/u/9lr15I330xOn04eemj8swET0xGsd99NRqPklVeGi8z36uzZ5NVXkw8+SF58cfnmAyZi9u8S/m3duuHU+5UryfPPJ2+//d9/4HfhQnL8eHL1anLqVPLyyxMdFVges31w9J/27Us+/zz5+efkvfeS114b7gZu2TKcrZqfT379dYjVqlXDa+CpU8nDD097cmBMenZYN28mGzYMhz8ff3z47Pvvh++1fvtteGV87LEhYBs3+kcG+B/qCVYyvBYeODB8pwWsOF3BWlxM1qyZ9hTAlHQFC1jROo41AESwgCKCBdQQLKCGYAE1BAuoIVhADcECaggWUEOwgBqCBdQQLKCGYAE1BAuoIVhADcECaggWUEOwgBqCBdQQLKCGYAE1BAuoIVhADcECaggWUEOwgBqCBdQQLKCGYAE1BAuoIVhADcECaggWUEOwgBqCBdQQLKCGYAE1BAuoIVhADcECaggWUEOwgBqCBdQQLKCGYAE1BAuoIVhADcECaggWUEOwgBqCBdQQLKCGYAE1BAuoIVhADcECaggWUEOwgBqCBdQQLKCGYAE1BAuoIVhADcECaggWUEOwgBqCBdQQLKCGYAE1BAuoIVhADcECaggWUEOwgBqCBdQQLKCGYAE1BAuoIVhADcECaggWUEOwgBqCBdQQLKCGYAE1BAuo8RePTsSh4r+O8AAAAABJRU5ErkJggg=="
/>
`;
